proposals/gc/MVP-JS.md

```diff
@@ -0,0 +1,538 @@
+
+To avoid problem, as a starting point, the following WebAssembly module:
+
+```wat
+;; example3.wat --> example3.wasm
+(module
+    (type $P (struct (field i32) (field i32)))
+    (type $R (struct (field $x1 (ref $P)) (field $x2 (ref $P))))
+    (func (export "goWild") (param (ref $R))
+        (struct.set $R $x1
+            (get_local 0)
+            (struct.new $P (i32.const 10) (i32.const 20)))
+    )
+)
+```
+
+will throw when passed a `Point` typed object because of the attempt to
+reinterpret the *nominal* field types of `Rect` with the *structural*
+field types of `$R`.
+
+```js
+WebAssembly.instantiateStreaming(fetch('example3.wasm'))
+.then(({instance}) => {
+    let r = new Rect(new Point(1,2), new Point(3,4));
+    instance.exports.goWild(r);  // throws at call boundary
+});
+```
+
+For the call to succeed, the WebAssembly module must intead *import* the
+exact Type Definition object and use that type import in all relevant
+function signatures:
+
+```wat
+;; example4.wat --> example4.wasm
+(module
+    (type $Point (import "" "Point") (eq (struct (field $x i32) (field $y i32))))
+    (type $Rect (struct (field $x1 (ref $Point)) (field $x2 (ref $Point))))
+    (func (export "goWild") (param (ref $Rect))
+        (struct.set $Rect $x1
+            (get_local 0)
+            (struct.new $Point (i32.const 10) (i32.const 20)))
+    )
+)
+```
+
+With this module, the following JS code can instantiate the module with a Type
+Definition object and then later pass it typed object instances of the same Type
+Definition object (but no other):
+
+```js
+function makeTypes() {
+    const Point = new StructType([{name: "x", type: int32}, {name: "y", type: int32}]);
+    const Rect = new StructType([{name: "p1", type: Point.ref}, {name: "p2", type: Point.ref}]);
+    return {Point, Rect};
+}
+
+const {Point, Rect} = makeTypes();
+
+WebAssembly.instantiateStreaming(fetch('example4.wasm'), {'':{Point}})
+.then(({instance}) => {
+    const r = new Rect(new Point(1,2), new Point(3,4));
+    instance.exports.goWild(r);  // succeeds
+    assert(r.x1 instanceof Point);
+    assert(r.x1.x === 10);
+    assert(r.x1.y === 20);
+
+    const {Point2:Point, Rect2:Rect} = makeTypes();
+    const r2 = new Rect2(new Point2(1,2), new Point2(3,4));
+    instance.exports.goWild(r2);  // throws at call boundary
+});
+```
+
+Thus, the expected nominal invariants of JS can be preserved by a combination of
+instantiation-time and run-time checks.
+
+But what about the other direction: what if WebAssembly exports a struct
+type definition that uses a `ref` field type? To reflect WebAssembly structural
+subtyping, we need new Value Type objects. Since the set of Value Type objects
+is open-ended, the WebAssembly JS API adds a new Value Type constructor
+function: `WebAssembly.ref(T)`. Using this, JS can create a suitable Type
+Definition object whose instances can be passed to the above `example3.wasm`
+because its fields are structurally typed:
```

Comment:
I don't quite understand this. Structural subtyping is a property of types, not fields.
Is it that case that `WebAssembly.ref(T)` constructs a JS Value Type that, when used as a field type, allows any structurally-compatible JS Value Type (and also WebAssembly type)? Afaict there can only be one such type (the principal type), so I assume that WebAssembly.ref(T) can actually return the same underlying value type for many such invocations (including WebAssembly.ref(F) that happens to be compatible?

Reply:
Great questions:
* You're right that structural subtyping is a property of types, not fields, and while `WebAssembly.ref(T)` is only demonstrated to be used for field types, Value Type objects can also be used (as described in the later section "Other WebAssembly uses of Value Type objects") globals, tables and (as mentioned in the TODO section) typed function params/results.
* You're right that `WebAssembly.ref(T)` is not generative and will return equivalent underlying value types for equivalent arguments.  This equivalence is demonstrated in the example where `makeTypes()` uses `WebAssembly.ref(Point)`, which is ultimately called twice, producing equivalent value types.